src/components/Avatar/Avatar.tsx

```diff
@@ -28,89 +30,136 @@ export interface Props {
   initials?: string;
   /** Whether the avatar is for a customer */
   customer?: boolean;
-  /** URL of the avatar image */
+  /** URL of the avatar image which falls back to initials if the image fails to load */
   source?: string;
   /** Accessible label for the avatar image */
   accessibilityLabel?: string;
 }
 
+export interface State {
+  hasError: boolean;
+  hasLoaded: boolean;
+  prevSource?: string;
+}
+
 export type CombinedProps = Props & WithAppProviderProps;
 
-function Avatar({
-  name,
-  source,
-  initials,
-  customer,
-  size = 'medium',
-  accessibilityLabel,
-  polaris: {intl},
-}: CombinedProps) {
-  const nameString = name || initials;
-
-  let finalSource: string | undefined;
-  let label: string | undefined;
-
-  if (accessibilityLabel) {
-    label = accessibilityLabel;
-  } else if (name) {
-    label = name;
-  } else if (initials) {
-    const splitInitials = initials.split('').join(' ');
-    label = intl.translate('Polaris.Avatar.labelWithInitials', {
-      initials: splitInitials,
-    });
-  } else {
-    label = intl.translate('Polaris.Avatar.label');
+export class Avatar extends React.PureComponent<CombinedProps, State> {
+  static getDerivedStateFromProps(props: Props, state: State) {
+    // store prevSource in state so we can compare when props change
+    if (props.source !== state.prevSource) {
+      return {
+        prevSource: props.source,
+        hasError: false,
+        hasLoaded: false,
+      };
+    }
+
+    return null;
   }
 
-  if (source) {
-    finalSource = source;
-  } else if (customer) {
-    finalSource = customerPlaceholder(nameString);
+  state: State = {
+    hasError: false,
+    hasLoaded: false,
```

Comment:
```suggestion
    hasLoaded: false,
    prevSource: this.props.source,
```

I think we’ll want to initialize `prevSource` in state

Reply:
This feels unnecessary, since it's never used after this initialization, and getDerivedStateFromProps is always used before render.